{% extends 'base.html' %}
{% load url from future %}
{% block title %}Gene {{ g.sys_name }}{% endblock %}
{% block javascript %}
	<link rel="stylesheet" href="/static/css/colorbox.css" />
    <script type="text/javascript" src="/static/javascripts/isblogo_mod_ab.js"></script>
    <script type='text/javascript' src='/static/javascripts/jquery.colorbox-min.js'></script>


<!-- 

GGBrowser scripts to be imported

-->
<link rel="stylesheet" href="/static/css/ggbrowserweb.css" />
<script type='text/javascript' src='/static/javascripts/ggbrowserweb/getData.js'></script>
<script type='text/javascript' src='/static/javascripts/ggbrowserweb/ggbrowserweb-functions.js'></script>

  
<script type="text/javascript" src="/static/javascripts/d3/d3.v3.min.js"></script>

<script src="/static/javascripts/js/select2.js"></script>

{% endblock %}
{% block nav %}
<!-- navigation starts-->
<div  id="nav">
	<ul>
    <li><a href="{% url 'index' %}">Home</a></li>
    <li id="current"><a href="{% url 'browse' %}">Browse</a></li>
    <li><a href="{% url 'search' %}">Search</a></li>
    <li><a href="{% url 'downloads' %}">Downloads</a></li>
    <li><a href="{% url 'contact' %}">Support</a></li>
    <li><a href="{% url 'about' %}">About</a></li>
	</ul>
	<!-- navigation ends-->
</div>	
{% endblock %}
{% block content %}
<!-- content-wrap starts -->
<div id="content-wrap">
	<div id="main">
    <script>
      $(document).ready(function() {
        
        $("#gres").dataTable({
          "bFilter": false,
          "bSort": true,
          "aaSorting": [[1, 'asc']],
          "bProcessing": true,
          "bServerSide": true,
          "sAjaxSource": "{% url 'gene_gres_json' %}{{species}}/{{gene}}"
        });
        $("#conditions").dataTable({
          "bFilter": false,
          "bSort": true,
          "aaSorting": [[2, 'asc']],
          "bProcessing": true,
          "bServerSide": true,
          "sAjaxSource": "{% url 'gene_conditions_json' %}{{species}}/{{gene}}"
        });
        $("#biclusters").dataTable({
          "bFilter": false,
          "bSort": true,
          "aaSorting": [[1, 'asc']],
          "bProcessing": true,
          "bServerSide": true,
          "sAjaxSource": "{% url 'gene_biclusters_json' %}{{species}}/{{gene}}"
        });
		$("#prom_arch").qtip({
			content: 'Promoter architecture calculated by scanning all CRMs clustered within a GRE across the entire genome. Only CRMs with MAST pvalue <= 1e-6 are shown. This was done to reveal potential TF-binding in unconventional locations (e.g. inside operons), which would not have been detected by cMonkey (see GRE table below)',
          	position: { target: 'mouse' },
          	show: 'click',
          	hide: { when: { event: 'click' } },           
		    style: 'dark' 
		   	});
		$("#gre_tip").qtip({
			content: 'Enriched GREs computed through biclusters by hypergeometric test with Benjamini-Hochberg corrected qvalue <= 0.05.',
          	position: { target: 'mouse' },
          	show: 'click',
          	hide: { when: { event: 'click' } },            
		    style: 'dark' 
		   	});
		$("#condition_tip").qtip({
			content: 'Enriched conditions computed through biclusters by hypergeometric test with Benjamini-Hochberg corrected qvalue <= 0.05.',
          	position: { target: 'mouse' },
          	show: 'click',
          	hide: { when: { event: 'click' } },            
		    style: 'dark' 
		   	});
		$("#bicluster_tip").qtip({
			content: 'Output of integrated biclustering algorithm, cMonkey (Reiss et al., 2006), that identifies groups of genes with (1) similar patterns of differential expression, over subsets of conditions (biclusters) (2) similar de novo detected sequence motifs in their promoters and (3) related functions, inferred from functional association networks (e.g., EMBL STRING (Szklarczyk et al., 2011)).',
          	position: { target: 'mouse' },
          	show: 'click',
          	hide: { when: { event: 'click' } },          
		    style: 'dark' 
		   	});
		$("#zoom_prom_arch").colorbox({  
			title:"{{ g.sys_name }} Promoter Architecture",
			});
      });
    </script>
		<table>
      <thead>
        <tr>
          <th class="first">Organism</th>
          <th>Gene name</th>
          <th>Common name</th>
          <th>Accession</th>
          <th>Description</th>
          <th>Start</th>
          <th>Stop</th>
          <th>Strand</th>
          <th>Chromosome</th>
        </tr>
      </thead>
      <tbody>
        <tr class="row-a">
          <td class="first"><a href="{% url 'species' %}{{ s.ncbi_taxonomy_id }}">{{ s.name }}</a></td>
          <td><a href="{% url 'genes' %}{{ s.ncbi_taxonomy_id }}/{{ g.sys_name }}">{{ g.sys_name }}</a></td>
          <td>{{ g.name }}</td>
          <td><a href="http://www.ncbi.nlm.nih.gov/protein/{{ g.accession }}">{{ g.accession }}</a></td>
          <td>{{ g.description }}</td>
          <td id="gene_start">{{ g.start }}</td>
          <td id="gene_end">{{ g.stop }}</td>
          <td>{{ g.strand }}</td>
          <td><a href="http://www.ncbi.nlm.nih.gov/nuccore/{{ g.chromosome.refseq }}">{{ g.chromosome.refseq }}</a></td>
        </tr>
      </tbody>
    </table>

    <blockquote><p>Corem membership</p></blockquote>
		<ul>
      {% for corem in corems %}
      <li><a href="{% url 'corems' %}{{ s.ncbi_taxonomy_id }}/{{ corem.corem_id }}">{{ corem.corem_id }}</a></li>
      {% endfor %}
    </ul>

<!-- GGBrowserweb starts-->

    <blockquote><p>Genome Browser</p></blockquote>
    <table id="table-ggbrowserweb">
      <td>
        <div id="boxes" class='select_div'>
        <div id="checkAll">
        <input type="checkbox" id="checkAll" value="checkAll" onchange="uncheckAll()" checked="true">Select All</input>
        </div></div>  
      </td>
      <td>
          <div id="ggbrowser-principal" style="background-color:white;width:778px;"></div>
      </td> 
      
      
      
      
    
      
    </table>

<script>
    //style="background-color:white;width:950px"  
    //#F9F9FF --> table-content-backgroung color      
          

//
var line = d3.svg.area()
  
      //.defined(function(d) { return d.START != null; })
      //.interpolate("basis")
      .x(function(d) { return x(d.START); })
      .y(function(d) {return yLine(d.MAX) ; })
  //.y0(390)
  //.y1(function(d) {return yLine(d.MAX) ; })
  //.attr("transform", "translate(0," + 250 + ")")
  ;

var genomeInfo = JSON.parse('{"id":189,"start":1,"name":"","type":"genes","end":202301,"left":0,"right":5000}');

 
  var margin = {top: 10, right: 30, bottom: 160, left: 40},
      margin2 = {top: 470, right: 30, bottom: 20, left: 40},
      margin3 = {top: 470, right: 30, bottom: 20, left: 40},
      width = 800 - margin.left - margin.right,
      height = 600 - margin.top - margin.bottom,
      height2 = 600 - margin2.top - margin2.bottom;

    
      height3 = 50;    
    
      var svg = d3.select("#ggbrowser-principal").append("svg")//#principal
        .attr("width", width + margin.left + margin.right)
        .attr("height", height + margin.top + margin.bottom)
        ;

    
    svg.append("defs").append("clipPath")
        .attr("id", "clip")
       
      .append("rect")
      //.attr("id", "rectClip")
        .attr("width", width)
        .attr("height", height)
        ;
        

  var focus = svg.append("g")
    .attr("transform", "translate(" + margin.left + "," + margin.top + ")");

  var context_segment = svg.append("g")
    .attr("transform", "translate(" + margin2.left + "," + margin2.top + ")");
                // I'll create a new path to fit context plot
              context_segment.append("defs").append("clipPath")
                      .attr("id", "clip2")
                    .append("rect")
                      .attr("width", width)
                      .attr("height", height2);

  var context = svg.append("g")
        .attr("id", "test")
        .attr("fill", "green")
        .attr("transform", "translate(" + margin2.left + "," + margin2.top + ")");
    
        context.append("rect")
        .attr("fill", "grey")
        .attr("opacity", 0.1)
            .attr("width", width)
      .attr("height", height);



  d3.select(".hand_cursor").style('border-style','dashed').style('border-color','grey');

  // Ajax Stuff - retrieving data from my ws
  //var id = 1;
  var view = {};
  var w = right - left
  //view.left = parseInt(d3.selectAll("#gene_start")[0][0].innerHTML) - 2500;
  //view.right = parseInt(d3.selectAll("#gene_end")[0][0].innerHTML) + 2500;

// read already populated info and use it to instanciate start/end fields
if((parseInt(d3.selectAll("#gene_start")[0][0].innerHTML) - 2500) <0) {
  view.left = 1;
  view.right = 5000;
  var left = 1;
  var right = 5000;
} 
else {
  view.left = parseInt(d3.selectAll("#gene_start")[0][0].innerHTML) - 2500;
  view.right = parseInt(d3.selectAll("#gene_end")[0][0].innerHTML) + 2500;
  var left = parseInt(d3.selectAll("#gene_start")[0][0].innerHTML) - 2500;
  var right = parseInt(d3.selectAll("#gene_end")[0][0].innerHTML) + 2500;
}

  var color = d3.scale.category20().domain(mot_names); //category10() 

  //var mot_names = ["MOT_1", "MOT_3", "MOT_4", "MOT_5", "MOT_10", "MOT_12", "MOT_16", "MOT_18", "MOT_30"];
  //drawCheckBoxAndGetMotData();
  //color for motif line and checkBox

var url_ = "{% url 'genes' %}{{ s.ncbi_taxonomy_id }}/"; // capturing url patern to make gene clickable 

  function drawBoxes(){
    

    //debugger;
  var box = d3.select("#boxes").selectAll("label")
  .data(mot_names);
  box.enter().append('label');
  box.exit().remove();
  box
      .attr('for',function(d,i){ return d ; })
      //.attr('color', "#ffff00")
      .text(function(d) { return d; })
      .style("color", function(d) { return color(d); } ) //
      .style( "text-anchor", "left" )
      .style( "font", "corrier")
      .style( "font-size", "16px")
  .append("input")
      .attr("id", "box_selected")
      .attr("name", "boxes")
      //.attr("checked", function(d) { debugger; eval(d + "checked");})
      /*.attr("checked", function(d) {
            if (typeof window[d + "checked"] != 'undefined'){
              return true;
            }
      else {
        return null;
      } 
      ;})*/
      .attr("checked", function(d) {
        if(window[d + "checked"] === false){
          return null;}
          else{
            return true;
          }})
      //.attr("checked", function(d) { if(d === 'pssm0') {return true;} else {return null ;};})
      .attr("type", "checkbox")
      .attr('color', "#ffff00")
      .style("fill", "#ffff00" )
      .attr("id", function(d,i) { return d; });
  // .append("br");
  //    .attr("onClick", changeMot);
  
  mot_names.forEach(function(d) {
    //delete(window[d + "checked"]);

    if (typeof window[d + "checked"] === 'undefined'){
      window[d + "checked"] = true;
    }
    /*else{
      if(window[d + "checked"] === false){
        console.log("exists : " +d)
        window[d + "checked"] = null;
      }
    }*/


  }
  );
  
  var check2 = d3.selectAll("input[name=boxes]").on("change", changed2);
  
  }




  var genetrack = {plus:10, minus:30, height:18};
  /*{plus:20, minus:60};//*/
  
  
  
  Array.prototype.contains = function(obj) {
      var i = this.length;
      while (i--) {
          if (this[i] === obj) {
              return true;
          }
      }
      return false;
  }



  function changePosition(text){
    
  getNameWithoutUnchecked().forEach(function(d) {
    svg.selectAll("#" + d).remove();
  
  } );  
  
    if(text == "plus" && genomeInfo.end >= view.right){
    
      view.left = left + 2500;
      view.right = right + 2500;
      x.domain([view.left, view.right]);
      x2.domain([view.left, view.right]);
      
      
      /*mot_names.forEach(function(d) {
      
        svg.select(".motifs_"+d).remove();
      
      });*/
        
      old_names = new Array();
      old_names = [];
      del = new Array();
      del = [];
      
      
      old_names = mot_names;  
        
      //getMotifData_all(view.left, view.right);
      //getMotifData_all_rank(view.right/w);
      
      
      //getListSegmentsAll(view.left,  view.right,view.right/w);
      //getSequenceAnnotationData(id, d3.round(view.left), d3.round(view.right), "gene");
      /*mot_names.forEach(function(d){
        //getOverlapMotifsGenes(d, "gene", id);
        getListSegments(d);
      });*/
      //getExpressionData(view.left, view.right);
      old_names.forEach(function(d) { 
        if(!mot_names.contains(d)){
          del.push(d);
          var t = svg.select(".motifs_"+d).remove();
          delete(t);
        }
        ;})
        
        cleanMemory(del);
        

    
      brush.clear();
      svg.select(".brush").call(brush);
    
      // updating the brush3 - genome
      //brush3.extent([view.left,view.right]);
      //svg2.select(".brush").call(brush3);
      left = view.left; right = view.right;
    
      focus.select(".x.axis").call(xAxis);
      context.select(".x.axis").call(xAxis2);




    }
    if(text == "minus" && genomeInfo.start < view.left){
      view.left = left - 2500;
      view.right = right - 2500;
      
      /*mot_names.forEach(function(d) {
      
        svg.select(".motifs_"+d).remove();
      
      }); */
        
        old_names = new Array();
        old_names = [];
        del = new Array();
        del = [];
      
      
        old_names = mot_names;  
        
        //getMotifData_all(view.left, view.right);
        //getMotifData_all_rank(view.right/w);
        
        //getListSegmentsAll(view.left,  view.right,view.right/w);
        //getSequenceAnnotationData(id, d3.round(view.left), d3.round(view.right), "gene");
        /*mot_names.forEach(function(d){
          //getOverlapMotifsGenes(d, "gene", id);
          getListSegments(d);
        }); */
        //getExpressionData(view.left, view.right);     
        old_names.forEach(function(d) { 
          if(!mot_names.contains(d)){
            del.push(d);
            svg.select(".motifs_"+d).remove();
          }
          ;}) 
        
        
      

      x.domain([view.left, view.right]);
      x2.domain([view.left, view.right]);
    
      // updating the brush3 - genome
      //brush3.extent([view.left,view.right]);
      //svg2.select(".brush").call(brush3);
      left = view.left; right = view.right;
    
    
      focus.select(".x.axis").call(xAxis);
      context.select(".x.axis").call(xAxis2);

    }

    
    //everytime I changeposition, I need to get new data, and bind it to chart as well (seqLogo)
    mot_names.forEach(function(d) {
      
      
      //getListSegments(d);
      //delete(window["_"+d]);
      window["_"+d] = focus.selectAll(".sequence-column" + d)
            //.data(eval(d));
      .data(eval("m_"+d))
      window["_"+d].remove().exit(); // ?
      window["_"+d]
            .enter()
            .append("g")
      .attr("class", "sequence-column")   ;
    });
    
    /*var tt = this.id;
    name_temp = [];
    mot_names.forEach(function(e) {
      //debugger;
      if(eval(e+"checked") != false){
        name_temp.push(e);
      };
    });
    max_local = getMaxLocal(name_temp);
    //debugger;*/
    
    redraw();
    //redrawExpressionAsPath();
    //redrawExpressionData();
    //drawCheckBoxAndGetMotData();
    drawLine();
    //brush3End();
    
    

/*    var e = document.getElementById("motif_class");
    var strUser = e.options[e.selectedIndex].value;
    console.log(strUser);
    changeMotifRepresentation(strUser);
*/    
    
    var check2 = d3.selectAll("input[name=boxes]").on("change", changed2);
  }

  
  function info(){
  
    var tr = d3.selectAll("#info")
      .data(genes.filter(function(d) { 
        return( (d[2] <= view.right) && (d[3] >= view.left)  );
      })).remove();      
          tr.enter().append("p")
          .attr("id", "info")
        .text(function(d) {return d[0]});
  
  
    }

  var padding = 20; // DO I NEED IT ?



  var x = d3.scale.linear().range([0, width]).domain([view.left, view.right]),
      x2 = d3.scale.linear().range([0, width]).domain([view.left, view.right]),
      x3 = d3.scale.linear().range([0, width]).domain([genomeInfo.start, genomeInfo.end]),
      y = d3.scale.linear().range([height, 0]),
      y2 = d3.scale.linear().range([height2, 0]).rangeRound([0, height2]);
      y3 = d3.scale.linear().range([40, 0]).rangeRound([0, 40]);
      //yMotif = d3.scale.linear().range([height/2, 0]);
      
  var ySeqLogo = d3.scale.linear()
    .range([( (height+margin.left)/2), 0]).domain([0, 2]);
  var normalize = d3.scale.linear()
    .range([0,2]);
      
      //yLine = d3.scale.linear().range([height2, 0]).domain([0,90]);
      var max_local = 50;
      yLine = d3.scale.linear().range([height, 100]).domain([0,max_local]); // 100 because I need to have room for other tracks
      
      yExpression = d3.scale.linear().range([height, 0]).domain([-5,5]);
      
  var xAxis = d3.svg.axis()
    .scale(x)
    .orient("bottom"),
      xAxis2 = d3.svg.axis().scale(x2).orient("bottom"),
      yAxis = d3.svg.axis().scale(y).orient("left");
      yAxisSeqLogo = d3.svg.axis().scale(ySeqLogo).orient("left").ticks(15);
      yAxisMaxLocal = d3.svg.axis().scale(yLine).orient("right").ticks(15);

    
  var xAxis3 = d3.svg.axis().scale(x3).orient("bottom");

    if ( (genomeInfo.end - genomeInfo.start) / 1000 > 1000){
      xAxis3
      .tickValues(d3.range(0, genomeInfo.end, 1000000))
      .tickFormat(function(d) {return (d/1000000) + 'M';});
    }
    else {
      xAxis3
      .tickValues(d3.range(10000, genomeInfo.end, 10000))
      .tickFormat(function(d) {return (d/1000) + 'k';});
    }
    
      //gene.attributes.split(";").forEach(function(d) {   var t = d.split("="); if(t[0]== "ID") console.log(t[1]);      })

  var brush = d3.svg.brush()
      .x(x2)
      //.on("brushstart", brushStart)
      .on("brush", brushed)
;
      //***********************.on("brushend", brushEnd);
              
  var brushFocus = d3.svg.brush()
      .x(x)
      .on("brush", brushedFocus)
      .on("brushstart", brushedFocusStart);
  //    .style('pointer-events', 'all');     


  /*var brush3 = d3.svg.brush()
      .x(x3)
      .on("brushend", brush3End)
      .on("brush", brushed3);   
*/




  
    var area = d3.svg.area()
        .x(function(d) { return x(d.START); })
        .y0(height)
        .y1(function(d) { return yLine(d.MAX)+300; });  
  

  var svg2 = d3.select("#allGenome").append("svg")//#principal
    .attr("id", "svg2")
          .attr("width", width + 50)
          .attr("height", height3)
  ;

      // some tests with checkbox and on
    
        var check = d3.selectAll("input[name=mot]").on("change", changed);
        //var checkBox = d3.select("input[value=\"grouped\"]").property("checked", changed)
        function changed(){
          if(this.checked){
          console.log("changed : " + this.value);
          svg.selectAll("#seqLogo")
          .style("display", null);
        
        }
        else {
          console.log(this)
          svg.selectAll("#seqLogo")
          .style("display", "none")
        }
        }
    


    
      //***********************
      // Reserved space for sequence logo vars etc. 
  
  
    var capHeightAdjust = 1.2;
    var xSeqLogo = d3.scale.ordinal()
      .rangeRoundBands([0, width], .1);//.domain([29451, 29465])

    


  // remove it 'cause I added the same code inside redrawSeqLogo; (then I update at once everytime I need to redraw)
  mot_names.forEach(function(d){
    //debugger;
    //debugger;
    window["_"+d] = focus.selectAll(".sequence-column" + d)
          .data(eval("m_"+d))
    //.data(eval(d))//["motif"]
          //.data(motif_all, function(e) { if (e.key === d) {return(e.values)}})
    .enter()
          .append("g")
    .attr("class", "sequence-column")   ;
    //debugger;
    ;
  //  delete(window[d]);
  //  console.log("called delete");
    })
  
    
  var yPos_ = 0;    
  mot_names.forEach(function(d){
    console.log('drawing Motif Heatmap');
  //  var yPos = 0;
    console.log('yPos : ' + yPos_);
    ////drawHeatmapMotif(d, eval("max_"+ d)[0].values, yPos);
    //drawOverlapMotif(d, eval("segment_"+ d)["MotifSegment"], yPos_, 0);
    //drawOverlapMotif(d, eval("segment_"+ d), yPos_, 0);
    yPos_ = yPos_ + (100/mot_names.length);
    console.log('yPos : ' + yPos_);
  });    
    


  var temp;
  //filtering my rect(genes) filtering only the view part of it


  //seqLogo

  /*var focus = svg.append("g")
      //.attr("transform", "translate(" +  0 + "," + margin.top + ")"); // margin.left instead of 0
      .attr("transform", "translate(" + margin.left + "," + margin.top + ")");
  var context = svg.append("g")
    //.attr("transform", "translate(" + 0 + "," + margin2.top + ")");
    .attr("transform", "translate(" + margin2.left + "," + margin2.top + ")");*/
  var genome = svg2.append("g")
          .attr("transform", "translate(" + 40 + "," + (0) + ")");


    genome.append("text")

  //  .attr("x", 10)
  //  .attr("y", 50)
    .attr("dx", 333) // padding-right
    .attr("dy", ".999em") // vertical-align: middle
    .attr("text-anchor", "center") // text-align: right;
    .attr("opacity", 0.5)
      .text(genomeInfo.name + " - " + genomeInfo.type + " - [" + genomeInfo.start + ":" + genomeInfo.end + "]");
      var b = 0;
      function filterData(dataSource, left, right){
      
        dd = dataSource.filter(function(d,i){  if(d[0].position >= left && d[0].position <= right) /*console.log(d.bits);*/ return d  ;});
  
  
    dd.forEach(function(d) {
          var y0 = 0;
      var b;
          d.bits = d.map( function( entry ) { 
        
            return  { position: +entry.position, bits: entry.bits, letter: entry.letter, y0: y0, y1 : y0 += +entry.bits };      
              }  
          )
          d.bitTotal = d.bits[d.bits.length - 1].y1; 
      //dataSource = d;
      //return b;
    
    });
  //  console.log(dataSource);
    return dd;
      }




        y.domain([0, 900]);
        x2.domain(x.domain());
        //y2.domain([1, 69]); //*** fix it, it may take max automaticaly
        
    focus.append("path")
          .attr("clip-path", "url(#clip)");
    
    focus.append("g")
            .attr("class", "x axis")
            .attr("transform", "translate(0," + height + ")")
            .call(xAxis);

        focus.append("g")
            .attr("class", "y axis")
        //.attr("transform", "translate(-1, 0)")
        .attr("transform", "translate(0, 195)") //--apaguei
            .call(yAxisSeqLogo);
        
        
        focus.append("g")
        .attr("class", "y axis")
        .attr("id", "max_local")
            .attr("transform", "translate(" + width + ",0)")
        .call(yAxisMaxLocal);
      
        
    context.append("g")
            .attr("class", "x axis")
            .attr("id", "brush1")
            .attr("transform", "translate(0," + (height2) + ")")
        //.attr("clip-path", "url(#clip2)")
            .call(xAxis2);


        context.append("g")
            .attr("class", "x brush")
        .attr("id", "t")
            .call(brush)
          .selectAll("rect")
            .attr("y", -6)
        //.attr("width", width - 40)
            .attr("height", height2+5)
        //.attr("clip-path", "url(#clip2)")
        ; 

            genome.append("g")
                .attr("class", "x axis")
          //.attr("id", "brush3")
                .attr("transform", "translate(0," + (height3 - padding) + ")")
                .call(xAxis3);

          /*genome.append("g")
              .attr("class", "x brush")
              .call(brush3)
            .selectAll("rect")
              .attr("y", -6)
              .attr("height", height3 + 7);
*/
  
getSequenceAnnotationData("{% url 'genes_json_annotation' %}{{species}}/",view.left);
getCre("/cres_in_range/{{species}}/",  view.left, view.right, 2);

init();


</script>


<!-- GGBrowserweb ends-->


    <blockquote><p>Promoter Architecture <a id="zoom_prom_arch" href="/static/images/promoter_architecture/{{ s.ncbi_taxonomy_id }}/{{ g.sys_name }}.png"> <img style="background:rgba(250, 250, 250, 0)" src="/static/zoom.png"></a> <a id="prom_arch" href="#" onclick="return false;"><img style="background:rgba(250, 250, 250, 0)" src="/static/help.png" ></a></p></blockquote>
    <img src="/static/images/promoter_architecture/{{ s.ncbi_taxonomy_id }}/{{ g.sys_name }}.png" background-color="black" width=100% height=100% class="float-center" />

    <blockquote><p>GREs <a id="gre_tip" href="#" onclick="return false;"><img style="background:rgba(250, 250, 250, 0); border:0px" src="/static/help.png" ></a></p></blockquote>
		<table id="gres" class="table">
      <thead>
        <tr>
          <th class="first">GRE ID</th>
          <th>qval</th>
          <th>PSSM</th>
        </tr>
      </thead>
      <tbody>
      </tbody>
    </table>
            
    <blockquote><p>Conditions <a id="condition_tip" href="#" onclick="return false;"><img style="background:rgba(250, 250, 250, 0); border:0px" src="/static/help.png" ></a></p></blockquote>
		<table id="conditions" class="table">
      <thead>
        <tr>
          <th class="first">Condition ID</th>
          <th>Condition Name</th>
          <th>qval</th>
        </tr>
      </thead>
      <tbody>
      </tbody>
    </table>
            
    <blockquote><p>Biclusters <a id="bicluster_tip" href="#" onclick="return false;"><img style="background:rgba(250, 250, 250, 0); border:0px" src="/static/help.png" ></a></p></blockquote>
		<table id="biclusters" class="table">
      <thead>
        <tr>
          <th class="first">Bicluster ID</th>
          <th>Residual</th>
          <th>Genes</th>
          <th>Conditions</th>
        </tr>
      </thead>
      <tbody>
      </tbody>
    </table>
		  <!-- main ends -->
		</div>
	
{% endblock %}
