{% extends 'base.html' %}
{% load url from future %}
{% block title %}EGRIN 2.0{% endblock %}
{% block nav %}
<!-- navigation starts-->	
	<div  id="nav">
		<ul>
			<li id="current"><a href="{% url 'index' %}">Home</a></li>
			<li><a href="{% url 'browse' %}">Browse</a></li>
			<li><a href="{% url 'search' %}">Search</a></li>
			<li><a href="{% url 'downloads' %}">Downloads</a></li>
			<li><a href="{% url 'contact' %}">Support</a></li>
			<li><a href="{% url 'about' %}">About</a></li>	
		</ul>
	<!-- navigation ends-->	
	</div>	
{% endblock %}
{% block content %}			
	<!-- content-wrap starts -->
	<div id="content-wrap">
	
		<div id="main">
		
		<script>
			$(document).ready(function() {
				$("#gre_tip").qtip({
					content: 'GRE: Gene Regulatory Element, a cluster of similar cis-regulatory motifs (CRMs) discovered in individual biclusters',
		          	position: { target: 'mouse' },
		          	show: 'click',
		          	hide: { when: { event: 'click' } },          
				    style: 'dark' 
				   	});
				$("#corem_tip").qtip({
					content: 'Corem: Co-regulated module, a set of conditionally co-regulated genes discovered by applying link community algorithm (Ahn et al., 2010) to backbone extracted (Serrano et al., 2009) gene-gene association network (inferred by cMonkey/Inferelator)',
		          	position: { target: 'mouse' },
		          	show: 'click',
		          	hide: { when: { event: 'click' } },          
				    style: 'dark' 
				   	});
				$("#bicluster_tip").qtip({
					content: 'Bicluster: Output of integrated biclustering algorithm, cMonkey (Reiss et al., 2006), that identifies groups of genes with (1) similar patterns of differential expression over subsets of conditions, (2) similar de novo detected sequence motifs in their promoters, and (3) related functions, inferred from functional association networks (e.g., EMBL STRING (Szklarczyk et al., 2011)).',
		          	position: { target: 'mouse' },
		          	show: 'click',
		          	hide: { when: { event: 'click' } },         
				    style: 'dark' 
				   	});
				$("#network_tip").qtip({
					content: 'Network: A full ensemble model consisting of (1) many (~100s) of individual cMonkey/Inferelator runs on blocked subsets of transcriptome data, (2) cis-regulatory motif (CRM) clustering, (3) genome-wide CRM scanning, and (4) corem detection.',
					position: { target: 'mouse' },
		          	show: 'click',
		          	hide: { when: { event: 'click' } },          
				    style: 'dark' 
				   	});
				$(function() {
    				$( "#accordion" ).accordion();
  					});
		      });
	    </script>
				
			<h3><strong>EGRIN 2.0 is a systems-level model that delineates the complex relationship between environment, gene regulation, and phenotype in prokaryotes</strong></h3>
            
			<a name="Why EGRIN 2.0?"></a>
			<h2><a href="index" title="">Why EGRIN 2.0?</a></h2>
					
            <p>A foremost challenge in systems biology is to understand how just a few transcription factors (TFs) in a microbial genome generate a wide array of nuanced responses to varied environmental challenges. <strong>EGRIN 2.0</strong> is a new model for the complete gene regulatory network (GRN) of a prokaryote. This model is reverse engineered directly from gene expression data and genomic sequence, and hence the methodology to generate EGRIN 2.0 is applicable to any prokaryotic organism.</p>

            <div id = "accordion">
            	<h3>More</h3>
            	<div>
            		<p>EGRIN 2.0 is <strong>remarkably accurate</strong>. It was able to discover nearly two-thirds of the <i>E. coli</i> TFs in RegulonDB <i>de novo</i>.</p>
					<p>In addition, EGRIN 2.0 describes how gene regulatory mechanisms and the environment integrate to coordinate cellular function, including:
						<ul>
							<li>The existence of gene regulatory elements (GREs) in both cannonical promoters and non-canonical locations, such as genes and operons</li>
							<li>How environment-specific TF influences at GREs reorganize functionally-related genes from different operons and regulons into conditionally co-regulated modules or <strong>corems</strong></li>
							<li>The highly similar fitness consequences of genes organized in corems</li>
						</ul>
					</p>
					<p>
					We have inferred comprehesive models for two prokaryotic organisms: a bacterium (<i><a href="http://egrin2.systemsbiology.net/networks/511145">Escherichia coli K-12</a></i>) and an archaeon (<i><a href="http://egrin2.systemsbiology.net/networks/64091">Halobacterium salinarum NRC-1</a></i>).
					</p>
            	</div>  
			</div>

			<p><code>This website facilitates <strong><a href="{% url 'search' %}">targeted</a></strong> and <strong><a href="{% url 'browse' %}">exploratory</a></strong> investigation of the model's predictions.</code></p>

            <h2>Highlights from the paper</h2>
            	<h3>Genes</h3>
            		<ul>
            			<li><a href="http://egrin2.systemsbiology.net/genes/511145/b0032"><i>E. coli</i> carA (b0032)</a></li>
            			<li><a href="http://egrin2.systemsbiology.net/genes/511145/b4246"><i>E. coli</i> pyrL (b4246)</a></li>
            			<li><a href="http://egrin2.systemsbiology.net/genes/64091/VNG6176G"><i>H. salinarum</i> kdpA (VNG6176G)</a></li>
            			<li><a href="http://egrin2.systemsbiology.net/genes/64091/VNG2349G"><i>H. salinarum</i> dppA (VNG2349G)</a></li>
            		</ul>
            	<h3>Corems</h3>
            		<ul>
            			<li><a href="http://egrin2.systemsbiology.net/corems/511145/ec516031"><i>E. coli</i> nucleotide corem (ec516031)</a></li>
            			<li><a href="http://egrin2.systemsbiology.net/corems/511145/ec516034"><i>E. coli</i> purine corem (ec516034)</a></li>
            			<li><a href="http://egrin2.systemsbiology.net/corems/511145/ec512157"><i>E. coli</i> pyrimidine corem (ec512157)</a></li>
            			<li><a href="http://egrin2.systemsbiology.net/corems/64091/hc21645"><i>H. salinarum</i> dpp corem (hc21645)</a></li>
            			<li><a href="http://egrin2.systemsbiology.net/corems/64091/hc6326"><i>H. salinarum</i> dpp leader corem (hc6326)</a></li>
            			<li><a href="http://egrin2.systemsbiology.net/corems/64091/hc21279"><i>H. salinarum</i> dpp permease corem (hc21279)</a></li>
            		</ul>
            <p></p>
			
			<!-- <blockquote><p>EGRIN 2.0 statistics</p></blockquote>
				
			<table>
				<tr>
					<th>value</th>
					<th>description</th>
				</tr>
                <tr class="row-a">
					<td><a href="{% url 'species' %}">{{ species }}</a></td>
					<td>Species modeled</td>
				</tr>
                <tr class="row-b">
					<td><a href="{% url 'networks' %}">{{ networks }}</a></td>
					<td>Networks inferred <a id="network_tip" href="#" onclick="return false;"><img style="background:rgba(250, 250, 250, 0); border:0px" src="/static/help.png" ></a></td>
				</tr>
                <tr class="row-a">
					<td><a href="{% url 'corems' %}">{{ corems }}</a></td>
					<td>Corems discovered <a id="corem_tip" href="#" onclick="return false;"><img style="background:rgba(250, 250, 250, 0); border:0px" src="/static/help.png" ></a></td>
				</tr>
				<tr class="row-b">
					<td><a href="{% url 'genes' %}">{{ genes }}</a></td>
					<td>Genes modeled</td>
				</tr>
				<tr class="row-b">
					<td><a href="{% url 'conditions' %}">{{ conditions }}</a></td>
					<td>Conditions in the models</td>
				</tr>
				<tr class="row-a">
					<td><a href="{% url 'gres' %}">{{ gres }}</a></td>
					<td>Gene Regulatory Elements (GREs) discovered <a id="gre_tip" href="#" onclick="return false;"><img style="background:rgba(250, 250, 250, 0); border:0px" src="/static/help.png" ></a></td>
				</tr>
                <tr class="row-b">
					<td><a href="{% url 'biclusters' %}">{{ biclusters }}</a></td>
					<td>Biclusters in the models <a id="bicluster_tip" href="#" onclick="return false;"><img style="background:rgba(250, 250, 250, 0); border:0px" src="/static/help.png" ></a></td>
				</tr>
                <tr class="row-a">
					<td><a href="{% url 'regulators' '511145' %}">{{ tfs }}</a></td>
					<td>RegulonDB annotated TFs mapped to GREs (E. coli only)</td>
				</tr>
			</table> -->
								
			<br />		
		<!-- main ends -->	
		</div>
        <div id="sidebar">
			
			<h3>Search</h3>	
			<form action="{% url 'search' %}" class="searchform">
				<p>
				<input name="search_query" class="textbox" type="text" />
  				<input name="search" class="button" value="Search" type="submit" />
				</p>			
			</form>	
			
            <h3>Updates</h3>
            <p>In development</p>
            <p class="post-info">Posted by <a href="http://baliga.systemsbiology.net/drupal/users/abrooks">Aaron</a></p><br/>
                    		
			<h3>Links Out</h3>
			<ul class="sidemenu">				
				<li><a href="https://www.systemsbiology.org/">Institute for Systems Biology</a></li>
                <li><a href="http://www.microbesonline.org/">MicrobesOnline</a></li>
            </ul>	
						
		<!-- sidebar ends -->		
		</div>		
{% endblock %}
        
